### PR TITLE
posix: implement fgetc()

### LINKED ISSUE #66936: posix: implement `fgetc()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`fgetc()` is required as part of `POSIX_DEVICE_IO` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/fgetc.html

### PR DESCRIPTION
**Pull request by osuchowski**
- Requires macros for getc() and getchar()
- Added documentation for fgetc, getc and getchar
- Created stub fgetc() in stdio.c

Closes #66936, #66943, #66944

### CODE REVIEW

**Comment by whythefyouflying** [comment]
Please bear with me as this is my first contribution. I am unsure how the minimal libc correlates with us having to create our own stdio headers and source files, should I create only tests for already existing libc implementations, or reimplement them? The `lib/libc/minimal/include/stdio.h` doesn't contain prototypes for `fgetc`, `getc` and `getchar`.

**Comment by cfriedt** [comment]
@whythefyouflying - it might be worth it to put this PR into draft mode

**Comment by whythefyouflying** [comment]
With the latest `6b104a2` commit I am getting a build error:
```
[86/171] Building C object zephyr/arch/arch/x86/core/CMakeFiles/arch__x86__core.dir/early_serial.c.obj
FAILED: zephyr/arch/arch/x86/core/CMakeFiles/arch__x86__core.dir/early_serial.c.obj 
ccache /home/dave/bin/zephyr-sdk-0.16.5/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=4096 -DPICOLIBC_LONG_LONG_PRINTF_SCANF -DTC_RUNID=cf8000df908bb5463ca9ac8ba808f239 -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/dave/zephyrproject/zephyr/kernel/include -I/home/dave/zephyrproject/zephyr/arch/x86/include -I/home/dave/zephyrproject/zephyr/include -I/home/dave/zephyrproject/zephyr/twister-out/qemu_x86/tests/posix/common/portability.posix.common.fs/zephyr/include/generated -I/home/dave/zephyrproject/zephyr/soc/x86/ia32 -I/home/dave/zephyrproject/zephyr/include/zephyr/posix -I/home/dave/zephyrproject/zephyr/lib/posix/options/getopt/. -I/home/dave/zephyrproject/zephyr/subsys/testsuite/include -I/home/dave/zephyrproject/zephyr/subsys/testsuite/ztest/include -I/home/dave/zephyrproject/zephyr/lib/posix -isystem /home/dave/zephyrproject/zephyr/lib/libc/common/include -m32  -Wshadow -fno-strict-aliasing -Werror -Os -imacros /home/dave/zephyrproject/zephyr/twister-out/qemu_x86/tests/posix/common/portability.posix.common.fs/zephyr/include/generated/autoconf.h -fno-printf-return-value -fno-common -g -gdwarf-4 -fdiagnostics-color=always -m32 -msoft-float -Wa,--divide --sysroot=/home/dave/bin/zephyr-sdk-0.16.5/x86_64-zephyr-elf/x86_64-zephyr-elf -imacros /home/dave/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/dave/zephyrproject/zephyr/tests/posix/common=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/dave/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/dave/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -mpreferred-stack-boundary=2 -mno-mmx -mno-sse --specs=picolibc.specs -D_POSIX_THREADS -std=c99 -MD -MT zephyr/arch/arch/x86/core/CMakeFiles/arch__x86__core.dir/early_serial.c.obj -MF zephyr/arch/arch/x86/core/CMakeFiles/arch__x86__core.dir/early_serial.c.obj.d -o zephyr/arch/arch/x86/core/CMakeFiles/arch__x86__core.dir/early_serial.c.obj -c /home/dave/zephyrproject/zephyr/arch/x86/core/early_serial.c
/home/dave/zephyrproject/zephyr/arch/x86/core/early_serial.c:43:2: error: #error "Unsupported configuration"
   43 | #error "Unsupported configuration"
      |  ^~~~~
/home/dave/zephyrproject/zephyr/arch/x86/core/early_serial.c: In function 'serout':
/home/dave/zephyrproject/zephyr/arch/x86/core/early_serial.c:72:17: error: implicit declaration of function 'IN'; did you mean 'MIN'? [-Werror=implicit-function-declaration]
   72 |         while ((IN(REG_LSR) & LSR_THRE) == 0) {
      |                 ^~
      |                 MIN
/home/dave/zephyrproject/zephyr/arch/x86/core/early_serial.c:74:9: error: implicit declaration of function 'OUT' [-Werror=implicit-function-declaration]
   74 |         OUT(REG_THR, c);
      |         ^~~
/home/dave/zephyrproject/zephyr/arch/x86/core/early_serial.c: In function 'z_x86_early_serial_init':
/home/dave/zephyrproject/zephyr/arch/x86/core/early_serial.c:100:21: error: 'mmio' undeclared (first use in this function)
  100 |         device_map(&mmio, X86_SOC_EARLY_SERIAL_MMIO8_ADDR, 0x1000, K_MEM_CACHE_NONE);
      |                     ^~~~
/home/dave/zephyrproject/zephyr/arch/x86/core/early_serial.c:100:21: note: each undeclared identifier is reported only once for each function it appears in
/home/dave/zephyrproject/zephyr/arch/x86/core/early_serial.c:100:27: error: 'X86_SOC_EARLY_SERIAL_MMIO8_ADDR' undeclared (first use in this function)
  100 |         device_map(&mmio, X86_SOC_EARLY_SERIAL_MMIO8_ADDR, 0x1000, K_MEM_CACHE_NONE);
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

**Comment by cfriedt** [comment]
@whythefyouflying - it might be necessary to do some console switching as well the way it is done in [tests/drivers/console_switching](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/console_switching). For example, you could probably use just about everything in that testsuite aside from `ZTEST(console_switching, test_write)`.

Note: the console is switched to the emulated uart before reading from the console (this is similar to `fgetc()`) and switched back to the normal uart before using `zassert_*()`. This is because the test subsystem needs to write to the console as well.

Switching to the emulated uart and back to the normal uart should almost be like a kind of critical section.

**Comment by whythefyouflying** [comment]
> @whythefyouflying - it might be necessary to do some console switching as well the way it is done in [tests/drivers/console_switching](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/console_switching). For example, you could probably use just about everything in that testsuite aside from `ZTEST(console_switching, test_write)`.
> 
> Note: the console is switched to the emulated uart before reading from the console (this is similar to `fgetc()`) and switched back to the normal uart before using `zassert_*()`. This is because the test subsystem needs to write to the console as well.
> 
> Switching to the emulated uart and back to the normal uart should almost be like a kind of critical section.

@cfriedt Unfortunately, for some reason the stdin is not connected with the emulated uart, not sure how to connect the two. I've uploaded my latest code, it's building but failing and the fgetc() call always gets a value of `-1` instead of the passed character.

**Comment by cfriedt** [comment]
> @cfriedt Unfortunately, for some reason the stdin is not connected with the emulated uart, not sure how to connect the two. I've uploaded my latest code, it's building but failing on the zassert for fgetc() call always gets a value of `-1` instead of the passed character.

I'll take a look when I can. Unfortunately, I'm a bit swamped right now.

File: tests/posix/headers/src/stdio_h.c [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#endif
+
+/**
+ * @brief Test existence and basic functionality of stdio.h
+ * 
+ * @see stdio.h
+*/
+ZTEST(posix_headers, test_stdio_h)
+{
+    #ifdef CONFIG_POSIX_API
```
Comment (cfriedt) [inline tests/posix/headers/src/stdio_h.c:20]:
```suggestion
```

File: tests/posix/headers/src/stdio_h.c [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    zassert_not_null((void *)getchar, "getchar is null");
+    zassert_not_null((void *)getc, "getc is null");
+    zassert_not_null((void *)fgetc, "fgetc is null");
+    #endif
```
Comment (cfriedt) [inline tests/posix/headers/src/stdio_h.c:24]:
```suggestion
```

File: tests/posix/stdio/boards/qemu_riscv64.overlay [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        status = "okay";
+    };
+
+    devmux1: dev_mux_1 {
+        compatible = "zephyr,devmux";
+        devices = <&uart0 &euart0 &euart1>;
+        zephyr,mutable;
+        selected = <2>;
+        status = "okay";
+    };
```
Comment (cfriedt) [inline tests/posix/stdio/boards/qemu_riscv64.overlay:38]:
```suggestion
```

File: tests/posix/stdio/boards/qemu_riscv64.overlay [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        current-speed = <0>;
+        status = "okay";
+    };
+
+    euart1: uart_emul1 {
+        compatible = "zephyr,uart-emul";
+        current-speed = <0>;
+        status = "okay";
+    };
+
```
Comment (cfriedt) [inline tests/posix/stdio/boards/qemu_riscv64.overlay:24]:
```suggestion
```

File: doc/services/portability/posix/option_groups/index.rst [hunk 5]
```diff
@@ -222,7 +222,7 @@ POSIX_DEVICE_IO
     feof(),
     ferror(),
     fflush(),
-    fgetc(),
+    fgetc(),yes (will fail with ``ENOSYS``:ref:`†<posix_undefined_behaviour>`)
```
Comment (cfriedt) [inline doc/services/portability/posix/option_groups/index.rst:225]: Technically, this shouldn't fail, as long as there is an open `FILE*`. But what happens when we do an `fopen()`? What path would we even pass to `fopen()` to get a `FILE*`? Is `stdin` already available? It should probably be (stdin, stdout, and stderr should be opened before `main()` in ISO C, but that is part of the c runtime).

That's just ISO C.

I think we are putting the cart before the horse so to speak.

The issue is that I don't believe we support file based streams (yet) in the libc / os level. We only support reading / writing to standard targets by integer file descriptor.

We might need to modify fdtable.c to allow us to use FILE based streams as well.

This seems to be a requirement for POSIX_DEVICE_IO in any case.

**Pull request closed without merging**